File: level1/lib/channels/pushover_channel.dart
import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart' show MediaType;

import '../core/channel.dart';
import '../core/models.dart';

class PushoverChannelSettings implements ChannelSettings {
  final String appToken;
  final String userKey;
  final String sound;
  final int priority;

  const PushoverChannelSettings({
    this.appToken = '',
    this.userKey = '',
    this.sound = '',
    this.priority = 0,
  });

  @override
  String get type => 'pushover';

  @override
  Map<String, dynamic> toJson() => {
        'appToken': appToken,
        'userKey': userKey,
        'sound': sound,
        'priority': priority,
      };

  @override
  List<String> get secretFields => ['appToken', 'userKey'];

  factory PushoverChannelSettings.fromJson(Map<String, dynamic> json) =>
      PushoverChannelSettings(
        appToken: json['appToken'] as String? ?? '',
        userKey: json['userKey'] as String? ?? '',
        sound: json['sound'] as String? ?? '',
        priority: json['priority'] as int? ?? 0,
      );
}

/// Sends alerts to Pushover via the messages.json endpoint. The app token and
/// user key are secrets carried in the request body/fields.
class PushoverChannel extends Channel {
  @override
  final String id;
  @override
  final bool enabled;
  @override
  final PushoverChannelSettings settings;
  final http.Client _client;

  static const _endpoint = 'https://api.pushover.net/1/messages.json';

  PushoverChannel({
    required this.id,
    this.enabled = true,
    required this.settings,
    http.Client? client,
  }) : _client = client ?? http.Client();

  @override
  String get type => 'pushover';

  Map<String, String> _fields(String message) => {
        'token': settings.appToken,
        'user': settings.userKey,
        'message': message,
        if (settings.sound.isNotEmpty) 'sound': settings.sound,
        'priority': settings.priority.toString(),
      };

  @override
  Future<void> send(AlertMessage message) async {
    final snapshot = message.snapshot;
    if (snapshot != null) {
      final request = http.MultipartRequest('POST', Uri.parse(_endpoint))
        ..fields.addAll(_fields(message.text))
        ..files.add(http.MultipartFile.fromBytes(
          'attachment',
          snapshot.bytes,
          filename: snapshot.name,
          contentType: MediaType.parse(snapshot.mimeType),
        ));
      final streamed = await _client.send(request);
      _check(await http.Response.fromStream(streamed));
    } else {
      final response = await _client.post(
        Uri.parse(_endpoint),
        headers: {'content-type': 'application/x-www-form-urlencoded'},
        body: _fields(message.text),
      );
      _check(response);
    }
  }

  @override
  Future<void> sendTest() async {
    await send(AlertMessage(
      timestamp: DateTime.now(),
      triggerType: 'test',
      text: 'Security Cam: test alert',
    ));
  }

  void _check(http.Response response) {
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw StateError(
          'Pushover failed (${response.statusCode}) ${response.body}');
    }
  }

  @override
  String? validate() {
    if (settings.appToken.isEmpty) return 'App token is required';
    if (settings.userKey.isEmpty) return 'User key is required';
    return null;
  }
}

File: level1/android/app/src/main/kotlin/io/securitycam/level1/channels/PushoverChannel.kt
package io.securitycam.level1.channels

import io.securitycam.level1.core.AlertMessage
import io.securitycam.level1.core.ChannelSettings
import java.util.concurrent.TimeUnit
import okhttp3.FormBody
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File

class PushoverChannelSettings(
    val appToken: String = "",
    val userKey: String = "",
    val sound: String = "",
    val priority: Int = 0,
) : ChannelSettings() {
    override val type: String get() = "pushover"
    override fun toJson(): Map<String, Any?> = mapOf(
        "appToken" to appToken,
        "userKey" to userKey,
        "sound" to sound,
        "priority" to priority,
    )
    override val secretFields: List<String> get() = listOf("appToken", "userKey")

    companion object {
        fun fromJson(json: Map<String, Any?>): PushoverChannelSettings = PushoverChannelSettings(
            appToken = json["appToken"] as? String ?: "",
            userKey = json["userKey"] as? String ?: "",
            sound = json["sound"] as? String ?: "",
            priority = (json["priority"] as? Number)?.toInt() ?: 0,
        )
    }
}

/**
 * Sends alerts to Pushover via the messages.json endpoint. The app token and
 * user key are secrets carried in the request body/fields.
 */
class PushoverChannel(
    override val id: String,
    override val enabled: Boolean = true,
    override val settings: PushoverChannelSettings,
    client: OkHttpClient? = null,
) : io.securitycam.level1.core.Channel {

    private val client: OkHttpClient =
        client ?: OkHttpClient.Builder().callTimeout(30, TimeUnit.SECONDS).build()

    override val type: String get() = "pushover"

    private fun fields(message: String): List<Pair<String, String>> = buildList {
        add("token" to settings.appToken)
        add("user" to settings.userKey)
        add("message" to message)
        if (settings.sound.isNotEmpty()) add("sound" to settings.sound)
        add("priority" to settings.priority.toString())
    }

    override suspend fun send(message: AlertMessage) {
        val snapshot = message.snapshot
        if (snapshot != null) {
            val tmp = File.createTempFile("level1", ".img")
            try {
                tmp.writeBytes(snapshot.bytes)
                val builder = MultipartBody.Builder()
                    .setType(MultipartBody.FORM)
                for ((k, v) in fields(message.text)) builder.addFormDataPart(k, v)
                builder.addFormDataPart(
                    "attachment",
                    snapshot.name,
                    tmp.asRequestBody(snapshot.mimeType.toMediaType()),
                )
                val response = client.newCall(Request.Builder().url(ENDPOINT).post(builder.build()).build()).execute()
                response.use { check(it.isSuccessful) { "Pushover failed (${it.code}) ${it.body?.string()}" } }
            } finally {
                tmp.delete()
            }
        } else {
            val body = FormBody.Builder().apply {
                for ((k, v) in fields(message.text)) add(k, v)
            }.build()
            val response = client.newCall(Request.Builder().url(ENDPOINT).post(body).build()).execute()
            response.use { check(it.isSuccessful) { "Pushover failed (${it.code}) ${it.body?.string()}" } }
        }
    }

    override suspend fun sendTest() {
        send(
            AlertMessage(
                timestamp = java.time.Instant.now(),
                triggerType = "test",
                text = "Security Cam: test alert",
            ),
        )
    }

    override fun validate(): String? {
        if (settings.appToken.isEmpty()) return "App token is required"
        if (settings.userKey.isEmpty()) return "User key is required"
        return null
    }

    companion object {
        private const val ENDPOINT = "https://api.pushover.net/1/messages.json"
    }
}